server/tso/tso.go

```diff
@@ -276,9 +303,11 @@ func (t *timestampOracle) getTS(leadership *election.Leadership, count uint32) (
 			log.Error("invalid timestamp", zap.Any("timestamp", current), errs.ZapError(errs.ErrInvalidTimestamp))
 			return pdpb.Timestamp{}, errs.ErrGenerateTimestamp.FastGenByArgs("timestamp in memory isn't initialized")
 		}
-
-		resp.Physical = current.physical.UnixNano() / int64(time.Millisecond)
-		resp.Logical = atomic.AddInt64(&current.logical, int64(count))
+		// Get a new TSO result with the given count
+		resp.Physical, resp.Logical = t.generateTSO(int64(count))
```

Comment:
Do you have any benchmark?

Reply:
Master version benchmark result:

> count:41504720, max:125, min:0, >1ms:26126877, >2ms:446177, >5ms:16044, >10ms:7130, >30ms:4008

This version benchmark result:

> count:47133934, max:134, min:0, >1ms:33838909, >2ms:1043347, >5ms:23782, >10ms:9214, >30ms:4036

It seems that the performances are almost the same.

Reply:
I think you should bench multiple clients, the `bench-tso-tools` only run one client.

Reply:
I modify the `pd-tso-bench` with 10 clients concurrently requesting. Here are the results with the same test duration.

Master version benchmark result (atomic):

> count:44143071, max:210, min:0, >1ms:13605, >2ms:202437, >5ms:8602550, >10ms:34522132, >30ms:799911

This version benchmark result (RWMutex):

> count:38797164, max:192, min:0, >1ms:9911, >2ms:154054, >5ms:6718519, >10ms:29477988, >30ms:2435771

It seems that we have little performance regression here. What do you think?

Reply:
I will test it a few more times.

Reply:
With 10 clients and each client has 1000 goroutines concurrently requesting.

`atomic` version benchmark results, 5 times, each lasts 1 minute:

```
count:22728286, max:236, min:0, >1ms:6422, >2ms:76352, >5ms:1536590, >10ms:14850813, >30ms:6256790
count:23579558, max:292, min:0, >1ms:4214, >2ms:70867, >5ms:1728370, >10ms:16573944, >30ms:5198453
count:22134959, max:232, min:0, >1ms:4841, >2ms:62373, >5ms:1387325, >10ms:14355729, >30ms:6323203
count:21140755, max:406, min:0, >1ms:3226, >2ms:74002, >5ms:1359579, >10ms:13158031, >30ms:6545016
count:22675841, max:211, min:0, >1ms:3759, >2ms:63362, >5ms:1463269, >10ms:15417254, >30ms:5726955
```

`RWMutex` version benchmark results, 5 times, each lasts 1 minute:

```
count:35658165, max:214, min:0, >1ms:8656, >2ms:118429, >5ms:3957681, >10ms:29724324, >30ms:1848697
count:21081062, max:237, min:0, >1ms:3871, >2ms:58648, >5ms:1230887, >10ms:12881581, >30ms:6905166
count:23707867, max:287, min:0, >1ms:8389, >2ms:80577, >5ms:1615269, >10ms:15939006, >30ms:6060615
count:24226733, max:254, min:0, >1ms:4100, >2ms:83970, >5ms:1854598, >10ms:16848481, >30ms:5435569
count:22391285, max:241, min:0, >1ms:5791, >2ms:83437, >5ms:1498255, >10ms:14303769, >30ms:6497415
```

Reply:
Plz update the test result into request description.

Reply:
I updated the PR description for the new benchmark results. PTAL @nolouch @Yisaer